daffodil-core/src/test/scala/org/apache/daffodil/general/TestRuntimeProperties.scala

```diff
@@ -0,0 +1,187 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.daffodil.general
+
+import org.junit.Test
+import org.junit.Assert.assertEquals
+import org.junit.Assert.assertFalse
+import org.junit.Assert.assertTrue
+
+import org.apache.daffodil.Implicits.intercept
+import org.apache.daffodil.dpath.NodeInfo
+import org.apache.daffodil.infoset.DISimple
+import org.apache.daffodil.infoset.ScalaXMLInfosetInputter
+import org.apache.daffodil.infoset.ScalaXMLInfosetOutputter
+import org.apache.daffodil.io.InputSourceDataInputStream
+import org.apache.daffodil.util.SchemaUtils
+import org.apache.daffodil.util.TestUtils
+
+/**
+ * Test InfosetOutputter that extends the ScalaXMLInfosetOutputter and redacts
+ * words specified in the dfdlx:runtimeProperties extension
```

Comment:
This redaction example is perhaps too simple. 

An interesting thing is when the special infoset outputter creates XML that is NOT valid w.r.t. the DFDL schema interpreted as an XML Schema.  One then needs a separate schema for validation from the one used for parsing/unparsing. 

If you add a pattern facet to the filtered element below in your DFDL schemas, such that things with the word REDACTED in them will NOT be valid w.r.t the pattern, that would be more representative of how this feature can change the XML in ways that violate the XML validity. 

This is tricky with respect to Daffodil's validation behaviors. "limited" validation will validate still, as it sees the infoset data during parsing. "full" validation should take the output of the infoset outputter and feed it to Xerces for validation, and that should create validation errors. 

I think tests which characterize these two different interactions of this feature with validation would be good additions.

Reply:
This raises an interesting issue/limitation.

The way Daffodil does full validation is it tee's the infoset events to two InfosetOutputters: one provided by the user (e.g. our redacting outputter) and the other a normal XMLTextInfosetOutputter. At the end of parsing, we run Xerces validation on the result of the XMLTextInfosetOutputter. This is necessary because Daffodil only knows how to validate the result of the XMLTextInfosetOutputter. If the user provided InfosetOutputter output json, NiFi records, some proprietary thing, etc. it wouldn't be able enable full validation otherwise.

And the runtimeProperty redaction/transformation only affects the output of the user provied InfosetOutputter. It does not modify the internal infoset, and it does not propagate any changes to the XMLTextInfosetOutputter used for full validation.

So Daffodil validation has no knowledge about what changes the user provided InfosetOutputter makes, and so enabling full validation change anything. Daffodil only validates the infoset as parsed.

We can think of InfosetOutputters as post-processing of the infoset, and validation occurs before that post-processing takes place. If something is expecting to validate the post-processed data make by a custom InfosetOutputter, it needs to do that in a different step outside of Daffodil.

Reply:
Ah, so in that case there's no need for the tests that address any interactions with validation, as there is none.